Update frontend dependencies

melck opened the pull request:
Upgrade frontend dependencies :
- Move external frontend dependencies to `puppetboard/static/libs`
- Move external frontend dependencies from `layout.html` to `_offline_static.html` and `_online_static.html`
- Upgrade fontrend dependencies
- Replacing jquery tablesort by datatable.net for all sortable tables
- Replacing c3.js by billboard.js and fix charts
- Fix loading buttons for datatables.net 
- Enabling jinja stip blocks to remove block from final HTML

Move to another PR after discussions
- Layout : 
  - Clean multiple block for loading javascript
  - Change content page on 15 column centered
  - Change menu to be fixed
- Trying to visually improve index page

@gdubicki This draft PR is from discussion on #705 with some visual changes. What do you think of it ?

gdubicki commented:
Hey @melck, BIG thanks for this contribution! It will greatly improve the maintainability of this project. 😊 🥳 🤝

I would be happy to test and review this PR as soon as it's ready.

However I do suggest to separate the changes to layout and possibly features (last 2 commits?) to yet another PR 😅 - it will be much faster to merge only the dependencies upgrade and code refactoring changes.

gdubicki commented:
Hey @melck! Do you need any help with this PR?

melck commented:
Hey @gdubicki, sorry I've been pretty busy, I'll take care of it before the weekend.

gdubicki commented:
> Hey @gdubicki, sorry I've been pretty busy, I'll take care of it before the weekend.

No problem, I don't want to rush you! I just don't want all your hard work on this to spoil. :)

melck commented:
Hello @gdubicki, i have updated this PR. I removed commits for changing layout and fixed unittest of offline statics.

Code coverage has reduced because of removing some unused code. I don't know why ...

gdubicki commented on puppetboard/templates/reports.html:
```diff
@@ -27,7 +27,7 @@
     {% endfor %}
```
Nit: I know that it's out of the scope of the PR, but it would be nice to unify this to look and work the same way as such filter in the `/nodes` page.

melck replied:
Yes i will do that later. When this PR is ok / merged, i will update my other one. After that i will open a new one for UI unification and improvements.  

To be honest, in the long run it would be easier to add a real frontend and only manage json returns on the backend. Remove all JS / CSS dependencies and start from scratch with, for example, the vuejs library and the element-plus framework.

Frontend update management would be simplified and optimized to handle large amounts of data

gdubicki replied:
I totally agree that the frontend of the app just needs to be replaced with some modern, standard solution. I don't have the skills to do that though, but if you are willing to implement it then it's great and you are welcome to do it! I will do my best to help you test it, review the code, get it merged and released. :)

bastelfreak replied:
I 100% agree with @gdubicki . I do not have the skills to write the frontend, but I think its a really good idea. I am happy to review and test PRs.

gdubicki replied:
Perhaps we should sync on Voxpupuli slack regarding this, @melck? If you want, please join it at https://puppetcommunity.slack.com/messages/voxpupuli/ . @bastelfreak and me are there under the same ids as here.

gdubicki commented on puppetboard/templates/report.html:
```diff
@@ -60,7 +37,7 @@ <h1>Summary</h1>
 <h1 id="logs">Logs</h1>
 {{ macros.checkbox_toggle("friendly", current_show_error_as, "friendly", "raw", "Friendly error messages") }}
-<table id="logs_table" class='ui basic compact table'>
+<table id="logs_table" class='ui very basic compact striped table'>
```
Nit: "striped" styling is really nice, makes reading a table so much easier! Perhaps you could apply this styling to all other tables, especially: nodes in overview, failures, nodes, facts, inventory and query result (in a table form)? Or the other way around: NOT apply it in this PR, but do apply it in a separate one?

melck replied:
I rollback this line. I think it would be better in a separate PR with UI unification.

gdubicki commented on puppetboard/templates/nodes.html:
```diff
@@ -1,34 +1,64 @@
+  data-paging="false"
+  data-info="false"
+  data-dom="t"
+  data-order='[[ 2, "asc" ]]'
```
The current default sort order is descending - from nodes with the newest reports to the oldest - and I think that we should keep it that way.

Can you change it, please?


Also, even if you change the order to descending, you still get the "unreported" nodes above the reported ones.

Example:

<img width="1326" alt="Screenshot 2022-08-29 at 17 18 21" src="https://user-images.githubusercontent.com/566632/187248227-aafc366d-10e5-47b4-907b-4ca287b52792.png">

This is also a different behavior than what we had and I don't think it's beneficial. (Firstly, it's not logical as the unreported nodes are not reported because they have been seen by the PuppetDB too long time ago. Secondly, it's just not practical as if you want to look at the nodes with the latest catalogs then these are not among them. Also their list may be a long one so it will take a lot of time to scroll through these.)

Can we change that too, please?

melck replied:
@gdubicki I made the change that you ask. I made a default to empty string for column ordering when no timestamp and i have done the same render behavior as Report column.

gdubicki reviewed (changes requested):
Thanks a lot for completing the PR, @melck! I did test it out and I noticed one thing that I think we have to fix and 2 others that we _might_ improve. Please check out my comments.

gdubicki commented:
@melck : I don't want to be a nuisance, but I still see a wrong order of nodes in index. :(

The code in the master:

<img width="1717" alt="Screenshot 2022-09-03 at 10 55 07" src="https://user-images.githubusercontent.com/566632/188265994-b41694d0-a8b4-44f3-8e8b-b1c096bac0cc.png">

v4.0.2:

<img width="1648" alt="Screenshot 2022-09-03 at 10 54 55" src="https://user-images.githubusercontent.com/566632/188266008-2adef944-ccb4-4dea-a00c-187023088fed.png">

It's not a matter of changing https://github.com/voxpupuli/puppetboard/blob/2b7500be83a59d33a82642c7aec5b15d3538b3d9/puppetboard/templates/index.html#L93-L93 to `desc` as then the unreported nodes are on the top, instead of being on the bottom...

gdubicki commented:
Also: I think you forgot to remove the striped style from the tables in nodes and report views, please see:
<img width="667" alt="Screenshot 2022-09-03 at 11 14 31" src="https://user-images.githubusercontent.com/566632/188266120-2e6fdbad-838f-4d6e-ad76-b9602b2491d6.png">

gdubicki commented:
@melck: sorry to bother you, but I hope that the above issues will be easier to fix for you than for me. And they are blocking the release.

melck commented:
Sorry i'll wheck that.  Do i have to creater another PR ?

gdubicki commented:
Yes, please create a new PR, @melck. Thank you!

gdubicki commented:
Thanks again @melck. The code from this PR has been pre-released in [4.1.0rc1](https://github.com/voxpupuli/puppetboard/releases/tag/v4.1.0rc1). Within a few days, after some testing at Egnyte (and hopefully other users) we will do a final 4.1.0 release.
